### DISCUSSION TITLE
Line Numbers in Editor move when adding new digit positions (going from 9 to 10 lines of code or 99 to 100)

### DISCUSSION THREAD

**Discussion by strangebirds** [godotengine/godot-proposals#7008] [Editor]
Godot version: 4.0.2 stable
System information: MacOS 12.6.6 (21G646) - Mac Studio M1 Max

Issue description:

In the GDScript editor the line numbers jump when adding a new digit position: i.e. when adding another line brings the line number from a single digit to two digits or from two to three digits. 9 to 10 or 99 to 100.

(I'll be honest, I didn't test 999 to 1000 or higher)

This is disconcerting to the user as they are focused on the code and the movement looks like an error appearing in the editor margin or text changing.

Expected behavior: the line numbers will not shift to accommodate more digits. The Controls for the editor window/view should be rendered with sufficient space for more digits (maybe four? 9999 lines of code seems unwieldy enough to be an upper limit)

Steps to reproduce

1. Start writing a new blank script
2. Exceed single digit numbers for the number of lines
3. Watch the line numbers jump to the right.
4. The line number movement can also be seen when removing a line takes it to a lower number of displayed digits: deleting line 10 causes all remaining line numbers (1-9) to shift left.

moved from issue [77786](https://github.com/godotengine/godot/issues/77786)

**Comment by strangebirds** [comment]
I think the user expects the editor to be a static environment unless something **important** happens (an error turns the line red, for example).

When the line number spacing it moves everything in the Editor window which makes me, at least, instinctively look for what changed: "what did I do wrong?"

@AThousandShips [brought up](https://github.com/godotengine/godot/issues/77786#issuecomment-1574328814) in the original issue that the gutter to the left of the editor space also holds breakpoints and bookmarking. I'm not sure I totally agree with this as an argument against since those other uses of space will always exist, regardless if more space is reserved for the line numbers. But it is a consideration to keep in mind.

**Reply by AThousandShips** [reply]
> those other uses of space will always exist

My point was that the area already takes up a fair bit of space without adding unused extra space like this

(Also tbh presenting it in this way, as someone else expressing something but that you don't agree, but that people should keep it in mind, when you relay someone else's opinion, is not very helpful. It is best to present all suggestions evenly and without injecting your own opinion)

**Reply by strangebirds** [reply]
Apologies for injecting my own opinion alongside your own.

**Comment by AThousandShips** [comment]
> Expected behavior: the line numbers will not shift to accommodate more digits. 

Can you provide an example of an editor that works this way please? Asked this already on the issue. Because "expected" kind of depends on this and if others will expect it too, in my experience this is *not* the way most IDEs work, and personal expectations are very individual, whereas what other software does is more of a point for not creating confusion for people

I can see this as a feature to *turn on* that is disabled *by default*, because I do not want this feature by default, or worse, impossible to turn off. And I can see plenty of people reporting this change as a **bug** if this gets turned on by default and can't be turned off

**Reply by strangebirds** [reply]
BBEDit (Mac), Notepad++ (Windows), Xcode (Mac) and Sublime Text (Mac and Windows) all reserve space for thousands positions and keep the numbers right-justified. JupyterLabs (online: https://jupyter.org/try-jupyter/lab/) reserves space for the tens position and begins shifting the editor windows when line 100 is reached. (I don't have a screenshot from Notepad++ or Xcode) These are the environments I have the most familiarity with.

<img width="210" alt="BBEdit" src="https://github.com/godotengine/godot-proposals/assets/9995188/2d4d41d7-51ec-48a5-93d1-87ddb2fcb790">
<img width="297" alt="SubLimeText" src="https://github.com/godotengine/godot-proposals/assets/9995188/e84f99d0-3654-4ca0-a1ef-9fa4977b0b22">
<img width="325" alt="JupterLabs" src="https://github.com/godotengine/godot-proposals/assets/9995188/7dae2dee-3ac7-4557-ae27-3df52b2dc277">

**Reply by AThousandShips** [reply]
Thank you for updating! (Visual Studio also has it apparently, though haven't tested, and there's at least some searches on google for "how do you remove this space")
